### DISCUSSION TITLE
Some cracks appear when I use smooth_shading

### DISCUSSION THREAD

**Discussion by stla** [pyvista/pyvista#1727] [Q&A] [answered]
Hello,

 I wrote a [marching cubes algorithm](https://gist.github.com/stla/384d33bb18d19154e23af94b28c00e85) (original code written in R in the **misc3d** package). I experimented it with the Barth sextic. I use spherical coordinates to clip the surface to a ball:

```python
phi = (1 + np.sqrt(5)) / 2
phi2 = phi * phi
def f(ρ, θ, ϕ):
    x = ρ*np.cos(θ)*np.sin(ϕ)
    y = ρ*np.sin(θ)*np.sin(ϕ)
    z = ρ*np.cos(ϕ)
    x2 = x * x
    y2 = y * y
    z2 = z * z
    return 4 * (phi2 * x2 - y2) * (phi2 * y2 - z2) * (phi2 * z2 - x2) - (1 + 2 * phi) * (x2 + y2 + z2 - 1) ** 2

mesh = marchingCubes(
    f, 0, 0, np.sqrt(3), 0, 2*np.pi, 0, np.pi, 150, 150, 150, rescale=True
)

def sph2cart(sph):  # spherical to Cartesian
    ρ = sph[0]
    θ = sph[1]
    ϕ = sph[2]
    return [
        ρ * cos(θ) * sin(ϕ),
        ρ * sin(θ) * sin(ϕ),
        ρ * cos(ϕ)
    ]
mesh.points = np.apply_along_axis(sph2cart, 1, mesh.points)
```

Here is the result when I don't use `smooth_shading`: 

![Barth_smoothFalse](https://user-images.githubusercontent.com/4466543/137551485-ddcea4c7-6664-4635-9401-0620ac81c2b8.png)

And with `smooth_shading = True`:

![Barth_smoothTrue](https://user-images.githubusercontent.com/4466543/137551553-722bb8ae-ccde-46c3-99f0-f95b39692ec4.png)

There are some cracks. Would you know where they come from, and how to get rid of them?

**Comment by stla** [comment, marked as answer]
Hmmm there's no crack when I increase the number of subdivisions of the voxel:

![BarthSextic_200_smooth](https://user-images.githubusercontent.com/4466543/137554424-b4813f21-b4e3-4660-abf2-6d0aeb4a5de1.png)

**Reply by akaszynski** [reply]
The "cracks" are actually a reversal or the orientation, or normals of the individual triangles.  Sometimes setting `auto_orient_normals=True` and using the `compute_normals` method can fix this, but in this case it doesn't work.

BTW, these are some awesome figures.  Mind if I reuse part of this as an example (or at least a sample mesh)?

**Reply by stla** [reply]
Hello @akaszynski ,

I've just done a [blog post](https://laustep.github.io/stlahblog/posts/MyPyVistaArtworks.html) with a selection of my PyVista animations, if you want to take a look :-)

**Reply by akaszynski** [reply]
They’re really amazing. We’re going to add to a link to this post to our external examples.  @tkoyama010 could you please add this when you get a chance?

**Reply by tkoyama010** [reply]
Of course. please see https://github.com/pyvista/pyvista/pull/2046.

**Reply by adeak** [reply]
@stla I'm not sure you'll get the ping on the pull request, so: are you happy with [this version](https://github.com/pyvista/pyvista/pull/2049/files)?

**Reply by stla** [reply]
@adeak Ye, that's nice. Thanks.

**Comment by stla** [comment]
Yes, I remarked this has to do with the orientation when I put on the culling. Feel free to use my images.

**Comment by stla** [comment]
If you are interested in my other PyVista artworks, you can check [this gist](https://gist.github.com/stla/189dbed406a023189e85a328603707c8), [this one](https://gist.github.com/stla/7fc562f3eb1704491aff69132b7221dd), and [my packages](https://github.com/stla?tab=repositories) whose name start with "Py".
I'm new to Python but I'm not new to 3D artwork. I worked with R (rgl package), OpenGL via Haskell, POV-Ray and threejs. PyVista is really great.
